### PR TITLE
Factor out server into a struct

### PR DESCRIPTION
**Pull request by will-lynas**

### CODE REVIEW

**Comment by will-lynas** [comment]
The git history is a bit of a mess. Will fix when #18 goes in.

**Review by LasterAlex — COMMENTED** [review]
This looks fine, you have done a lot of changes i had to do a long time ago!
For some reason phrase_bot is broken with latest-teloxide branch though..

File: teloxide_tests/src/mock_bot.rs [hunk 1]
```diff
@@ -70,63 +76,60 @@ fn find_chat_id(value: Value) -> Option<i64> {
...
-    /// Caught responses from the server
-    pub responses: Mutex<Option<Responses>>,
-    /// Stack size used for dispatching
-    pub stack_size: Mutex<usize>,
-    bot_lock: Mutex<Option<MutexGuard<'static, ()>>>, // Maybe in the future ill make something like an atomic
-                                                      // bool that says if the bot is locked or not, and implement a custom Drop trait
+    pub dependencies: DependencyMap,
+
+    current_update_id: AtomicI32,
+    stack_size: usize,
```
Comment (LasterAlex) [inline teloxide_tests/src/mock_bot.rs:127]: `.stack_size(&mut self, stack_size: usize)` is needed, or at least the field needs to be public

**Review by LasterAlex — COMMENTED** [review]
Awesome! I am very happy to see that the only global mutex is the bot lock! I'll add a few commits to fix some things

**Comment by LasterAlex** [comment]
After the conversations are resolved I'll merge it, all the changes that this commit needed to do were done
I'll make another pr adding error_handler, distribution_function and others, while you can try to add a parallelized feature
We'll see about MockBotBuilder after all of that
Thanks for improving the crate!

File: teloxide_tests/src/mock_bot.rs (hunk 1 above)
Comment (will-lynas) [reply teloxide_tests/src/mock_bot.rs:128]: Oops, yeah that should've been public

File: teloxide_tests/src/mock_bot.rs [hunk 2]
```diff
@@ -25,8 +28,9 @@ use teloxide::{
     types::Me,
 };
 
-static GET_POTENTIAL_STORAGE_LOCK: Mutex<()> = Mutex::new(());
-static BOT_LOCK: Mutex<()> = Mutex::new(());
+lazy_static! {
+    static ref BOT_LOCK: Arc<Mutex<()>> = Arc::new(Mutex::new(()));
+}
```
[the remark this answers is not in the document]
Comment (will-lynas) [reply teloxide_tests/src/mock_bot.rs:36]: No idea why I had that

**Pull request merged**